flink/v2.1/flink/src/main/java/org/apache/iceberg/flink/data/StructRowData.java

```diff
@@ -186,8 +186,31 @@ private BigDecimal getDecimalInternal(int pos) {
 
   @Override
   public TimestampData getTimestamp(int pos, int precision) {
-    long timeLong = getLong(pos);
-    return TimestampData.fromEpochMillis(timeLong / 1000, (int) (timeLong % 1000) * 1000);
+    if (precision == 9) {
+      // Nanosecond precision: get nanoseconds directly from struct
+      Object longVal = struct.get(pos, Object.class);
+      long nanos;
+      if (longVal instanceof Long) {
+        nanos = (long) longVal;
+      } else if (longVal instanceof OffsetDateTime) {
+        nanos = Duration.between(Instant.EPOCH, (OffsetDateTime) longVal).toNanos();
+      } else if (longVal instanceof LocalDateTime) {
+        nanos =
+            Duration.between(Instant.EPOCH, ((LocalDateTime) longVal).atOffset(ZoneOffset.UTC))
+                .toNanos();
+      } else {
+        throw new IllegalStateException(
+            "Unknown type for nanosecond timestamp field. Type name: "
+                + longVal.getClass().getName());
+      }
+      long mills = Math.floorDiv(nanos, 1_000_000);
+      int leftover = Math.floorMod(nanos, 1_000_000);
+      return TimestampData.fromEpochMillis(mills, leftover);
```

Comment:
Question: Why not something similar to the line below with `/`, `%` and `*` operators?